libzfs_core/_error_translation.py

```diff
@@ -0,0 +1,469 @@
+            raise lzc_exc.NameInvalid(fromsnap)
+        elif not _is_valid_snap_name(snapname):
+            raise lzc_exc.NameInvalid(snapname)
+        elif fromsnap is not None and len(fromsnap) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(fromsnap)
+        elif len(snapname) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(snapname)
+        elif fromsnap is not None and _pool_name(fromsnap) != _pool_name(snapname):
+            raise lzc_exc.PoolsDiffer(snapname)
+    elif ret == errno.ENOENT and fromsnap is not None:
+        if not _is_valid_snap_name(fromsnap):
+            raise lzc_exc.NameInvalid(fromsnap)
+    if ret == errno.ENOENT:
+        raise lzc_exc.SnapshotNotFound(snapname)
+    raise _generic_exception(ret, snapname, "Failed to estimate backup stream size")
+
+
+def lzc_receive_xlate_error(ret, snapname, fd, force, origin, props):
+    if ret == 0:
+        return
+    if ret == errno.EINVAL:
+        if not _is_valid_snap_name(snapname):
+            raise lzc_exc.NameInvalid(snapname)
+        elif len(snapname) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(snapname)
+        elif origin is not None and not _is_valid_snap_name(origin):
+            raise lzc_exc.NameInvalid(origin)
+        else:
+            raise lzc_exc.BadStream()
+    if ret == errno.ENOENT:
+        if not _is_valid_snap_name(snapname):
+            raise lzc_exc.NameInvalid(snapname)
+        else:
+            raise lzc_exc.DatasetNotFound(snapname)
+    if ret == errno.EEXIST:
+        raise lzc_exc.DatasetExists(snapname)
+    if ret == errno.ENOTSUP:
+        raise lzc_exc.StreamFeatureNotSupported()
+    if ret == errno.ENODEV:
+        raise lzc_exc.StreamMismatch(_fs_name(snapname))
+    if ret == errno.ETXTBSY:
+        raise lzc_exc.DestinationModified(_fs_name(snapname))
+    if ret == errno.EBUSY:
+        raise lzc_exc.DatasetBusy(_fs_name(snapname))
+    if ret == errno.ENOSPC:
+        raise lzc_exc.NoSpace(_fs_name(snapname))
+    if ret == errno.EDQUOT:
+        raise lzc_exc.QuotaExceeded(_fs_name(snapname))
+    if ret == errno.ENAMETOOLONG:
+        raise lzc_exc.NameTooLong(snapname)
+    if ret == errno.EROFS:
+        raise lzc_exc.ReadOnlyPool(_pool_name(snapname))
+    if ret == errno.EAGAIN:
+        raise lzc_exc.SuspendedPool(_pool_name(snapname))
+
+    raise lzc_exc.StreamIOError(ret)
+
+
+def _handleErrList(ret, errlist, names, exception, mapper):
+    if ret == 0:
+        return
+
+    if len(errlist) == 0:
+        suppressed_count = 0
+        if len(names) == 1:
+            name = names[0]
+        else:
+            name = None
+        errors = [mapper(ret, name)]
+    else:
+        errors = []
+        suppressed_count = errlist.pop('N_MORE_ERRORS', 0)
+        for name, err in errlist.iteritems():
+            errors.append(mapper(err, name))
+
+    raise exception(errors, suppressed_count)
+
+
+def _pool_name(name):
+    return re.split('[/@#]', name, 1)[0]
```

Comment:
`/` is hiarchical name component separator, `@` is name-snapshot separator, what's `#`?  Possibly worth a comment here explaining these three magic characters (bytes?).